plugins/modules/ec2_vpc_route_table.py

```diff
@@ -494,45 +543,94 @@ def ensure_subnet_association(connection=None, module=None, vpc_id=None, route_t
+            return True
         new_association_ids.append(result['association_id'])
 
     if purge_subnets:
-        to_delete = [a_id for a_id in current_association_ids
-                     if a_id not in new_association_ids]
-
-        for a_id in to_delete:
+        to_delete = [association_id for association_id in current_association_ids
+                     if association_id not in new_association_ids]
+        for association_id in to_delete:
             changed = True
-            if not check_mode:
+            if not module.check_mode:
                 try:
-                    connection.disassociate_route_table(aws_retry=True, AssociationId=a_id)
+                    connection.disassociate_route_table(aws_retry=True, AssociationId=association_id)
                 except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
                     module.fail_json_aws(e, msg="Couldn't disassociate subnet from route table")
 
-    return {'changed': changed}
+    return changed
+
+
+def disassociate_gateway(connection, module, route_table):
+    # Delete all gateway associations that have state = associated
+    # Subnet associations are handled in its method
+    changed = False
+    associations_to_delete = [association['RouteTableAssociationId'] for association in route_table['Associations'] if not association['Main']
+                              and association.get('GatewayId')
+                              and (association['AssociationState']['State'] == 'associated' or association['AssociationState']['State'] == 'associating')]
+    for association_id in associations_to_delete:
+        changed = True
+        if not module.check_mode:
+            try:
+                connection.disassociate_route_table(aws_retry=True, AssociationId=association_id)
+            except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+                module.fail_json_aws(e, msg="Couldn't disassociate gateway from route table")
+
+    return changed
+
 
+def associate_gateway(connection, module, route_table, gateway_id):
+    filters = ansible_dict_to_boto3_filter_list({'association.gateway-id': gateway_id, 'vpc-id': route_table['VpcId']})
+    try:
+        route_tables = describe_route_tables_with_backoff(connection, Filters=filters)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Couldn't get route tables")
+    for table in route_tables:
+        if table.get('RouteTableId'):
+            for association in table.get('Associations'):
+                if association['Main']:
+                    continue
+                if association.get('GatewayId') and association['GatewayId'] == gateway_id and (association['AssociationState']['State'] == 'associated'
+                                                                                                or association['AssociationState']['State'] == 'associating'):
```

Comment:
Minor cosmetic suggestion, you could simplify slightly:
```suggestion
                if association.get('GatewayId', '') and (association['AssociationState']['State'] in ['associated', 'associating']):
```